[docs] Update README

abiogenesis-now opened the pull request:
@hausdorff @jessicayuen Would appreciate eyes and feedback on this even thought it's not 100% done. 

**TODOs:**
- [X] ~Figure out why `kuard` image from the quickstart script isn't actually showing anything (something to do with the port forwarding?)~ (swapped out with guestbook image)
- [X] Update README in `docs/cli-reference` to be more helpful (this is temporary until we figure out how to use this autogenerated markdown on the site)

Let me know if you have any questions. Hopefully the short annotations in the *Additional Documentation* help give a sense of where this is going.

jessicayuen commented on CONTRIBUTING.md:
```diff
@@ -1,64 +1,58 @@
-## Managing vendor/
```
Hmm, I don't think we should be removing the contents of the original file, which I found particularly helpful.

Maybe add the DCO Sign off as another section to this file?

hausdorff replied:
Just so we're all clear: we actually shouldn't remove the contributing file, as we need it for scripts.

jessicayuen commented on CODE-OF-CONDUCT.md:
```diff
@@ -0,0 +1,37 @@
+This code of conduct applies both within project spaces and in public spaces
+when an individual is representing the project or its community.
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be reported by contacting a project maintainer, Alex Clemmer (alex@heptio.com) and TBD.
```
I think it's sufficient to follow the Ark/Sonobouy model and mention `by contacting the project maintainer(s)`, so we don't have to constantly update this file.

Or - replace with the ksonnet / heptio ksonnet mailing list.

jessicayuen commented on README.md:
```diff
@@ -6,75 +6,149 @@
+### Prerequisites
+* *You should have access to an up-and-running Kubernetes cluster (**maximum** version 1.7).* If you do not have a cluster, [choose a setup solution](https://kubernetes.io/docs/setup/) from the official Kubernetes docs.
+
+* *You should have `kubectl` installed.* If not, follow the instructions for [installing via Homebrew (MacOS)](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-with-homebrew-on-macos) or [building the binary (Linux)](https://kubernetes.io/docs/tasks/tools/install-kubectl/#tabset-1)
```
Missing `.` at end of line.

jessicayuen commented on README.md:
````diff
@@ -6,75 +6,149 @@
+
+In one shell session:
+```
+# Start by creating your app directory'
````
Accidental `'`?

jessicayuen commented on README.md:
```diff
@@ -6,75 +6,149 @@
+
+* **[CLI Reference](/docs/cli-reference)** - What ksonnet commands are available, and how do I use them?
+
+* **Concept Reference(WIP)** - What do all these special ksonnet terms mean (e.g. *prototypes*) ?
```
space before `(`

jessicayuen commented:
Do you think you can squash some of these commits before merging? Ex: address comments

abiogenesis-now commented:
yeah, I've been changing a fair amount, will rebase once they're reviewed---
@jessicayuen, do you mind taking a look at the two most recent commits? (`address comments` and `update...`)

TLDR of what's changed:
* `CONTRIBUTING.md`
* changed quickstart script to use guestbook image instead of kuard image (https://heptio.slack.com/archives/C49LH42KD/p1510844821000483)
* added some stuff to the CLI readme (linked from main README)

If you don't mind, I'll make a followup PR to add the `concepts.md` stuff, just so we can get the updated README merged first. let me know if you have any questions, thanks!

jessicayuen commented on README.md:
```diff
@@ -6,75 +6,168 @@
+Here we provide a shell script that shows some basic ksonnet features in action. You can run this script to deploy and update a basic web app UI, via a Kubernetes Service and Deployment. This app is shown below:
+
+<p align="center">
+<img alt="guestbook screenshot" src="/docs/img/guestbook.png" style="width:60% !important;">
```
Nit: Can we remove the "af", "blah", etc. words from this screenshot? Or update it to "Hello, ksonnet" or something.

abiogenesis-now replied:
good catch

hausdorff replied:
lol I kind of like "af", but if we want to be all "professional" I understand.

jessicayuen commented on README.md:
````diff
@@ -6,75 +6,168 @@
+Copy and paste the script below to deploy the container image for a basic web app UI:
+
+```
+# Start by creating your app directory
````
I do think it's worth mentioning here that we're initializing the app based on the current context in their configurations.

abiogenesis-now replied:
good point

jessicayuen commented on README.md:
```diff
@@ -6,75 +6,168 @@
+# Deploy your manifest to your cluster
+ksonnet apply default
+
+# Set up an API proxy so that you can access the guestbook service locally
```
What's your opinion on removing this entirely and asking them to visit their server? I feel that they should have this information based on the pre-reqs.

abiogenesis-now replied:
I want to avoid blocking users who aren't familiar with services/bastion hosts/etc. `LoadBalancer` doesn't work on all clusters (only AWS/GCP/Azure, *and* it's billable), so I opted for API proxy as a relatively agnostic way of getting them to the webpage

jessicayuen commented on README.md:
````diff
@@ -6,75 +6,168 @@
+ks show default
+```
+
+If you're wondering how ksonnet differs from existing tools, the [full-length tutorial (WIP)]() shows you how to use other ksonnet features to implement the rest of the Guestbook app (so that the buttons work!).
````
The full-length tutorial goes to a missing page. Is this intentional or is there supposed to be a stub (i.e. to a TODO page similar to concepts)?

abiogenesis-now replied:
yes, the tutorial will be hosted on the new ksonnet site, and is not yet completed. if it is easier (and less confusing), I can remove the link for now

jessicayuen commented on docs/concepts.md:
```diff
@@ -0,0 +1,24 @@
+### Prototype
+TODO
+
+### Parameter
```
I know you're going to be updating this in a separate PR, but just wanted to mention that `Environment` should probably be one of these concepts.

abiogenesis-now replied:
ahh right, thanks!

abiogenesis-now commented:
@jessicayuen I addressed most of your comments (and left a response on the one I didn't change). PTAL, thank you!

hausdorff commented on CONTRIBUTING.md:
````diff
@@ -1,64 +1,96 @@
 ```
-# See unused
-govendor list +unused
+make all
````
Are we calling out a dependency on `GOPATH` here? I must be missing it.

hausdorff commented on CONTRIBUTING.md:
````diff
@@ -1,64 +1,96 @@
+```
+brew install coreutils
 ```
+Running `make` again afterwards should work.
````
Might want to call out `make test` here too if this is the contributing doc.

hausdorff commented on CONTRIBUTING.md:
```diff
@@ -1,64 +1,96 @@
+### Manage dependencies
 
-### Add a new dependency
+This project uses `govendor` to manage Go dependencies. To make things easier for reviewers, put updates to `vendor/` in a separate commit within the same PR.
```
I'd specifically call out the requirements of adding a vendored dependency. Perhaps something like (this is a really quick cut, feel free to discard):

```
This project uses `govendor` to manage dependencies. When you introduce a dependency to the ksonnet codebase, you need to:

1. Discuss the use case and need for the dependency with the maintainers by opening an issue.
1. After building agreement, vendor the dependency using `govendor`.
  1. Where possible, **please pin the repository to a specific stable version.** Please do not vendor HEAD if you can avoid it!
  1. **Please introduce a vendored dependency in a dedicated commit to make review easier!**
1. Write the code that exercises the dependency.
```

hausdorff commented on CONTRIBUTING.md:
````diff
@@ -1,64 +1,96 @@
-% govendor fetch -v +missing
+# Download missing dependencies into vendor/
+govendor fetch -v +missing
 ```
````
Following from the above comment, I'd also mention that you should check in the changes to `vendor` separately from your new app code.

hausdorff commented on CONTRIBUTING.md:
```diff
@@ -1,64 +1,96 @@
+   tag=vX.Y.Z
+   git fetch   # update
+   git tag -s -m $tag $tag origin/master
+   git push origin tag $tag
```
I believe this is handled if you do something like `VERSION=0.7 make`?

hausdorff commented on README.md:
```diff
@@ -6,75 +6,169 @@
 
 *STATUS: Development is ongoing—this tool is pre-alpha.*
 
 ## Install
```
Probably should make a note to add a "releases" section when we cut a release, too.

abiogenesis-now replied:
^ this is in the *Changelog* section at the bottom of the README

abiogenesis-now commented:
right it should still be there

On Thu, Nov 16, 2017 at 3:35 PM Alex Clemmer <notifications@github.com>
wrote:

> *@hausdorff* commented on this pull request.
> ------------------------------
>
> In CONTRIBUTING.md
> <https://github.com/ksonnet/ksonnet/pull/99#discussion_r151529886>:
>
> > @@ -1,64 +1,58 @@
> -## Managing vendor/
>
> Just so we're all clear: we actually shouldn't remove the contributing
> file, as we need it for scripts.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ksonnet/ksonnet/pull/99#discussion_r151529886>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AYOZMjJ_S2JK0BDxvon53XdjpdPAAJ0Xks5s3Jx-gaJpZM4Qfx7B>
> .
>

hausdorff commented on README.md:
```diff
@@ -6,75 +6,169 @@
 ## Install
 
-To build from source:
+> You should have Go installed, and an appropriately set `$GOPATH`. (For most systems, the minimum Go version is 1.7. However, recent OSX may [require golang >=1.8.1](https://github.com/golang/go/issues/19734) to
```
We support only > 1.8 actually. We're using stuff like `sort.Slice`.

hausdorff commented on README.md:
```diff
@@ -6,75 +6,169 @@
+> You should have Go installed, and an appropriately set `$GOPATH`. (For most systems, the minimum Go version is 1.7. However, recent OSX may [require golang >=1.8.1](https://github.com/golang/go/issues/19734) to
+avoid an immediate `Killed: 9`.) If you need additional help, see the [official Go installation guide](https://golang.org/doc/install#install).
+
+ksonnet is installed like any other Go binary. Run the following shell commands to download ksonnet and ensure that it is runnable in any directory:
```
How about something like: `ks` is installed like any other Go binary.

hausdorff commented on README.md:
````diff
@@ -6,75 +6,169 @@
+PATH=$PATH:$GOPATH/bin
+```
+
+If your ksonnet is properly installed, you should be able to run the following `--help` command and see similar output:
````
Might want to standardize on calling the binary `ks` to avoid confusion.

hausdorff commented on README.md:
```diff
@@ -6,75 +6,169 @@
+
+# Autogenerate a basic manifest
+ksonnet generate deployment-exposed-with-service guestbook-ui \
+--name guestbook \
```
Might want to consider indenting these flags. Like:

```
ksonnet generate deployment-exposed-with-service guestbook-ui \
  --name guestbook \
  --image alpinejay/dns-single-redis-guestbook:0.3 \
  --type ClusterIP
```

abiogenesis-now commented:
@hausdorff addressed all your comments, PTAL when you get a chance, thanks!

abiogenesis-now commented:
^ left it in a separate commit for now so changes are easier for you to review
